apps/sms/services/js/service_manager.js

```diff
@@ -0,0 +1,239 @@
+    ['moz-mobile-connections', MozMobileConnectionsServiceInitializer],
+    ['moz-settings', MozSettingsServiceInitializer]
+  ]);
+
+  function initializeShims(endpoint, upgradeKey) {
+    if (endpoint.MozMobileMessageShim) {
+      endpoint.MozMobileMessageShim.init(
+        endpoint.navigator.mozMobileMessage,
+        upgradeKey ? new endpoint.BroadcastChannel(
+          `${endpoint.MozMobileMessageShim.name}-${upgradeKey}`
+        ) : endpoint
+      );
+
+      debug('Initialized MozMobileMessageShim (upgrade key: %s)', upgradeKey);
+    }
+
+    if (endpoint.MozSettingsShim) {
+      endpoint.MozSettingsShim.init(
+        endpoint.navigator.mozSettings,
+        upgradeKey ? new endpoint.BroadcastChannel(
+          `${endpoint.MozSettingsShim.name}-${upgradeKey}`
+        ) : endpoint
+      );
+
+      debug('Initialized MozSettingsShim (upgrade key: %s)', upgradeKey);
+    }
+
+    if (endpoint.MozMobileConnectionsShim) {
+      endpoint.MozMobileConnectionsShim.init(
+        endpoint.navigator.mozMobileConnections,
+        upgradeKey ? new endpoint.BroadcastChannel(
+          `${endpoint.MozMobileConnectionsShim.name}-${upgradeKey}`
+        ) : endpoint
+      );
+
+      debug(
+        'Initialized MozMobileConnectionsShim (upgrade key: %s)',
+        upgradeKey
+      );
+    }
+  }
+
+  var shimHostPromise;
+  function getShimEndpoint(upgradeKey) {
+    debug('Request shim endpoint');
+
+    if (shimHostPromise) {
+      debug('Shim endpoint is ready');
+      return shimHostPromise;
+    }
+
+    if (upgradeKey) {
+      debug('Shim endpoint exists, but is not cached yet');
+      shimHostPromise = Promise.resolve(window.open('', upgradeKey));
+    } else {
+      debug('Shim endpoint does not exist yet');
+      shimHostPromise = new Promise((resolve) => {
+        var shimHost = window.open(
+          '/views/shared/shim_host.html', App.instanceId, 'alwaysLowered'
+        );
+
+        shimHost.addEventListener('load', function onLoad() {
+          shimHost.removeEventListener('load', onLoad);
+          resolve(shimHost);
+
+          debug('New shim endpoint is created and cached');
+        });
+      });
+    }
+
+    return shimHostPromise;
+  }
+
+  var ServiceManager = {
+    [priv.upgradeKey]: null,
+
+    initialize(serviceKeys) {
+      var upgradeKey = this[priv.upgradeKey] =
+        sessionStorage.getItem(UPGRADE_KEY) || null;
+
+      debug('Upgrade key: %s', upgradeKey);
+
+      if (!upgradeKey) {
+        initializeShims(exports, upgradeKey);
+      }
+
+      serviceKeys.forEach((serviceKey) => {
+        serviceInitializers.get(serviceKey).initialize(upgradeKey);
+      });
+    },
+
+    upgrade(serviceKeys) {
+      debug('Upgrade is requested');
+
+      if (this[priv.upgradeKey]) {
+        return Promise.resolve();
+      }
+
+      var upgradeKey = App.instanceId;
+
+      return getShimEndpoint().then((endpoint) => {
+        initializeShims(endpoint, upgradeKey);
```

Comment:
> Not sure if it's a proper timing for shim init because we run upgrade when fully loaded.

We run "upgrade" on ```visuallyLoaded```. :)

> And, dose conversation service getAllConversations method really work for the if we init the mobile message shim here?

Yep, it woks :)

> Some api like settings.setReadAheadThreadRetrieval might be needed before fully loaded, but we can definitely change the timing of calling API. 

Sorry I had to explain general idea in advance, it's not that easy to see what is going on looking at the code only :/

So the idea is the following:
* We have _all_ necessary clients/services in window on startup; In this case all shims and clients listen on ```window``` only (no ```BC``` or ```MC``` involved);
* On visually loaded we run "upgrade" process:
  * Create alwaysLowered window - in this case we'll have copies of all shims, but they listen on ```BC``` - so no conflict here with existing in-window shims, nobody uses these services yet;
  * Once we have new shims ready (inside alwaysLowered window), we start to upgrade existing window clients:
    * Destroy window clients: with the help of my brand-new finalize (not the best name though) bridge client plugin, every client has ```finalize``` method that returns promise that is resolved once all pending operations (method calls, streams) are finished and client is disconnected from the corresponding service;
    * Immediately create "upgraded" client (either ```SharedWorker``` based like ```ConversationClient``` or alwaysLowered based like ```MozMobileConnectionsClient```). So at this step, we have two clients simultaneously: old that's waiting for pending operations to complete, and the new "upgraded" one that will handle all next requests.
    * Since we can have pending operations (we run upgrade on visuallyLoaded so ```getAllConversations``` stream is still active), we wait for those operations to complete _and only then_ we destroy/disconnect window client and corresponding window based service that was backing window client.
* Now all new requests are handled by "upgraded" clients.

So all this not-that-easy machinery and its temporary and workaround-y nature made me think that having this in single file is a good idea, but I maybe completely wrong :)